src/uiHelpers_nanos.c

```diff
@@ -0,0 +1,131 @@
+        { \
+            BAGL_LABELINE,ID_HEADLESS, 0, 12, 128, \
+            12,0,0,0,0xFFFFFF,0, \
+            BAGL_FONT_OPEN_SANS_REGULAR_11px|BAGL_FONT_ALIGNMENT_LEFT,0 \
+        }, \
+        "HEADLESS ",0,0,0,NULL,NULL,NULL \
+    }
+#endif
+
+enum {
+    ID_UNSPECIFIED = 0x00,
+    ID_ICON_GO_LEFT = 0x01,
+    ID_ICON_GO_RIGHT = 0x02,
+    ID_ICON_CONFIRM = 0x03,
+    ID_ICON_REJECT = 0x04,
+
+    ID_HEADLESS = 0xff,
+};
+
+static const bagl_element_t ui_busy[] = {
+    UI_BACKGROUND(),
+    UI_TEXT(ID_UNSPECIFIED, 0, 20, 128, "..."),
+};
+
+static unsigned int ui_busy_button(
+        unsigned int button_mask MARK_UNUSED,
+        unsigned int button_mask_counter MARK_UNUSED
+)
+{
+    return 0;
+}
+
+static const bagl_element_t ui_scrollingText[] = {
+    UI_BACKGROUND(),
+    UI_ICON_LEFT(ID_ICON_GO_LEFT, BAGL_GLYPH_ICON_LEFT),
+    UI_ICON_RIGHT(ID_ICON_GO_RIGHT, BAGL_GLYPH_ICON_RIGHT),
+
+    // TODO(ppershing): what are the following magical numbers?
+
+    UI_TEXT(ID_UNSPECIFIED, 0, 12, 128, &displayState.scrolling.header),
+    UI_TEXT(ID_UNSPECIFIED, 0, 26, 128, &displayState.scrolling.currentText),
+    #ifdef HEADLESS
+    HEADLESS_UI_ELEMENT(),
+    #endif
+};
+
+static const bagl_element_t* ui_prepro_scrollingText(const bagl_element_t *element)
+{
+    scrollingState_t* ctx = scrollingState;
+    ASSERT(ctx->initMagic == INIT_MAGIC_SCROLLER);
+
+    switch (element->component.userid) {
+    case ID_ICON_GO_LEFT:
+        return (ctx->scrollIndex == 0)
+               ? NULL
+               : element;
+    case ID_ICON_GO_RIGHT:
+        return (ctx->scrollIndex + SIZEOF(ctx->currentText)
+                >= strlen(ctx->fullText) + 1)
+               ? NULL
+               : element;
+    default:
+        // Always display all other elements.
+        return element;
+    }
+}
+
+static void uiCallback_init(ui_callback_t* cb, ui_callback_fn_t* confirm, ui_callback_fn_t* reject)
+{
+    cb->state = CALLBACK_NOT_RUN;
+    cb->confirm = confirm;
+    cb->reject = reject;
+}
+
+static const bagl_element_t ui_confirm[] = {
+    UI_BACKGROUND(),
+    UI_ICON_LEFT(ID_ICON_REJECT, BAGL_GLYPH_ICON_CROSS),
+    UI_ICON_RIGHT(ID_ICON_CONFIRM, BAGL_GLYPH_ICON_CHECK),
+    UI_TEXT(ID_UNSPECIFIED, 0, 12, 128, displayState.confirm.header),
+    UI_TEXT(ID_UNSPECIFIED, 0, 26, 128, displayState.confirm.text),
+    #ifdef HEADLESS
+    HEADLESS_UI_ELEMENT(),
+    #endif
+};
+
+static const bagl_element_t* ui_prepro_confirm(const bagl_element_t *element)
+{
+    confirmState_t* ctx = confirmState;
+
+    ASSERT(ctx->initMagic == INIT_MAGIC_CONFIRM);
+    switch (element->component.userid) {
+    case ID_ICON_REJECT:
+        return ctx->callback.reject ? element : NULL;
+    case ID_ICON_CONFIRM:
+        return ctx->callback.confirm ? element : NULL;
+    default:
+        // Always display all other elements.
+        return element;
+    }
+}
+
+void ui_displayBusy()
+{
+    UX_DISPLAY(ui_busy, NULL);
+}
+
+void ui_displayConfirm_init()
```

Comment:
I would call it maybe `run` and not `init` ? `init` sounds like it should be called before we set up display data in our contexts